### PR TITLE
Added System.Index and System.Range

### PR DESCRIPTION
**Pull request by 333fred**
Implementations are copied from corefx, with minor changes around the
formatting methods, throw helpers, and HashCode to remove unavailable
apis.

You can find corefx's versions here:
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Index.cs
https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/Range.cs

### CODE REVIEW

File: src/Compilers/Core/Portable/InternalUtilities/Index.cs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        /// <param name="fromEnd">Indicating if the index is from the start or from the end.</param>
+        /// <remarks>
+        /// If the Index constructed from the end, index value 1 means pointing at the last element and index value 0 means pointing at beyond last element.
+        /// </remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Index(int value, bool fromEnd = false)
+        {
+            if (value < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(value), value, "Non-negative number required.");
```
Comment (sharwell) [inline src/Compilers/Core/Portable/InternalUtilities/Index.cs]: 💡 These exceptions should be localized if we intend to use the types.

**Comment by 333fred** [comment]
> 💡 Can you link these into Microsoft.CodeAnalysis.Workspaces at the same time?

I intentionally did not, and I think we need to be a lot more intentional and think about the ramifications if we were to expose these types in that layer. No one has IVT to the compiler except our unit tests, meaning that if we end up moving to a version of the framework that has these types built-in, we can remove our classes without having to think about any ramifications on partner teams. If we want to add the Index/Range classes to higher layers in roslyn, I think that _needs_ to come with an analyzer that will fail the build if the types are used in a type signature, even internal ones, until we manage to remove all IVTs to the code base.

File: src/Compilers/Core/Portable/InternalUtilities/Index.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/Core/Portable/InternalUtilities/Index.cs]: Not unless the messages in `ExceptionUtilities` also need to be localized. These types are intentionally internal, and consumers cannot directly create one of these. If this exception is thrown, _we_ have a bug.

**Comment by sharwell** [comment]
> If we want to add the Index/Range classes to higher layers in roslyn, I think that needs to come with an analyzer that will fail the build if the types are used in a type signature, even internal ones, until we manage to remove all IVTs to the code base.

This is not really the problem it sounds like it is from this post, for two reasons:

1. We are not adding new IVTs, and the ones already in place are getting progressively removed
2. We are not blocked from removing the types because if a problem occurred we would just use `[TypeForwardedTo]` from the workspaces layer.

I'm fine with not adding it in this pull request, but I would not be surprised to see it added by someone in a different pull request shortly afterward. It's mostly about whether it gets added once or twice.

**Comment by CyrusNajmabadi** [comment]
> No one has IVT to the compiler except our unit tests, 

I don't think this is a request for IVT.  More a request to literally just add them as a linked file.  It's something we do a *lot* between Compiler/Workspace in order to share a lot of completely fine utility code (i.e. collections, pooling, etc.).  

> if we end up moving to a version of the framework that has these types built-in, we can remove our classes without having to think about any ramifications on partner teams. 

If you end up moving to a version of the framework that has these built in, it already has ramifications for the Workspace :)   And it would be super weird for compiler to change their framework and not have the workspace do that too.

And, if that is something that needs to happen, the solution is simple.  Just remove these from compiler layer, but keep them in workspace layer.

**Comment by 333fred** [comment]
> I don't think this is a request for IVT. More a request to literally just add them as a linked file. It's something we do a lot between Compiler/Workspace in order to share a lot of completely fine utility code (i.e. collections, pooling, etc.).

I know what Sam is asking for here. I'm concerned by doing that, as there C# language features that depend on these types, and the framework specifically did not do the work with these types to ensure that type definitions unify correctly when defined in multiple places. Other internal teams have IVT to the workspace layer, which adds a few different complications, particularly if they've already done a similar thing of defining Index/Range themselves. If they haven't, it means that they could silently start depending on our CodeAnalysis binaries for language features as well. If the IDE layer wants to think through the ramifications of introducing these types at the workspace layer and introduce them in a separate PR, I'm fine with it, but I'm not comfortable doing it in this PR.

**Comment by 333fred** [comment]
> If you end up moving to a version of the framework that has these built in, it already has ramifications for the Workspace :) And it would be super weird for compiler to change their framework and not have the workspace do that too.

I'm not worried about the workspace here. I'm worried about teams with IVT to the workspace.

**Comment by CyrusNajmabadi** [comment]
> I'm not worried about the workspace here. I'm worried about teams with IVT to the workspace.

I see.  Thanks for the clarification! :)

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler please review.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review please.

**Pull request merged**